common/src/main/java/io/pravega/common/concurrent/AsyncSemaphore.java

```diff
@@ -0,0 +1,220 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.concurrent;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.Exceptions;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * A synchronization primitive that allows executing arbitrary (concurrent) tasks, where each task requires a well-known
+ * number of credits, subject to a total number of credits being available. Each task's successful execution will "borrow"
+ * its share of credits, and a task cannot execute if the number of credits available is insufficient. Credits may be
+ * restored externally using the {@link #release} method.
+ *
+ * This is similar to {@link java.util.concurrent.Semaphore}, except that this class allows for asynchronous processing
+ * and each task can request an arbitrary numbe of credits. It can be useful in solving problems making use of the
+ * Leaky Bucket Algorithm (https://en.wikipedia.org/wiki/Leaky_bucket).
+ */
+@ThreadSafe
+public class AsyncSemaphore implements AutoCloseable {
+    //region Members
+
+    private final int totalCredits;
+    @GuardedBy("queue")
+    private int usedCredits;
+    @GuardedBy("queue")
+    private final ArrayDeque<DelayedTask> queue;
+    @GuardedBy("queue")
+    private boolean closed;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the {@link AsyncSemaphore} class.
+     *
+     * @param totalCredits Total number of available credits.
+     * @param usedCredits  Initial number of used credits.
+     */
+    public AsyncSemaphore(int totalCredits, int usedCredits) {
+        Preconditions.checkArgument(totalCredits > 0, "totalCredits must be a positive integer");
+        Preconditions.checkArgument(usedCredits >= 0, "usedCredits must be a non-negative integer");
```

Comment:
The condition is for it to be non-negative. Zero is allowed and is usually the case for new segments.

Reply:
A caller can only acquire credits if they want to execute a task.

I need this constructor with initial used credits so that I can initialize this post-recovery. Post-recovery there may be unindexed entries that still need processing, whose number is stored in a one of the Segment's attributes. I don't see why I need to go through the async `run` method just to initialize a number.